resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/DriverServiceFeatureStep.scala

```diff
@@ -38,15 +38,21 @@ private[spark] class DriverServiceFeatureStep(
     s"$DRIVER_HOST_KEY is not supported in Kubernetes mode, as the driver's hostname will be " +
       "managed via a Kubernetes service.")
 
+  def isDigit(char: Char) : Boolean = {
+    char >= '0' && char <= '9'
+  }
+
   private val preferredServiceName = s"${kubernetesConf.resourceNamePrefix}$DRIVER_SVC_POSTFIX"
-  private val resolvedServiceName = if (preferredServiceName.length <= MAX_SERVICE_NAME_LENGTH) {
+  private val resolvedServiceName = if (preferredServiceName.length <= MAX_SERVICE_NAME_LENGTH
+    && !isDigit(preferredServiceName.charAt(0))) {
```

Comment:
This is still wrong. You are saying the first character needs to be a-z. This only checks if it's a digit.

Reply:
!isDigit(preferredServiceName.charAt(0))   ：
if true :  that means the first character is not digit；
if false : That means the first character is digit.

If the first character is digit ,  see the code `else{...}`.
If the first character is not digit, we don't need to deal with the name of Service.

Reply:
You're not seeing my point. I understand what this code does.
The title of this PR says that you must check if the first character is a-z. That is not the same as being a digit.

Reply:
I want to express :
The name of Service needs to satisfies the regular expression. Otherwise, we need to deal with it further.

Reply:
Look at the first part of the regular expression you are trying to match. It does not say "not a digit" which is what you are checking. You need to check whether it's a-z.

Reply:
I got it. You mean  the first character is not only digit such as other symbols.

Reply:
I say digit,  because resourceNamePrefix  that has been processed and contains three kinds :
(1)digit
(2)a-z
(3)- 


```
  def getResourceNamePrefix(appName: String): String = {
    val id = KubernetesUtils.uniqueID()
    s"$appName-$id"
      .trim
      .toLowerCase(Locale.ROOT)
      .replaceAll("\\s+", "-")
      .replaceAll("\\.", "-")
      .replaceAll("[^a-z0-9\\-]", "")
      .replaceAll("-+", "-")
  }
```

Reply:
Sure, but why not make the logic consistent with the desired outcome here? it's just simpler and catches the '-' case.

Reply:
I think only Service name needs  to satisfy that the first character is a-z. The Pod name can pass the validation for pod's regex when use resourceNamePrefix.

Reply:
This change is still wrong then. You are not checking whether the first char is a-z.

Reply:
You are right. I add a regex `replaceAll("^-", "")`.
Now ,the method `getResourceNamePrefix ` returns  value that the first character is 0-9 or a-z.